web/src/message_view.js

```diff
@@ -661,6 +661,98 @@ export function show(raw_terms, opts) {
     }
 }
 
+function handle_anchor_case({
+    anchor,
+    fetch_status_shows_anchor_fetched,
+    message_list_data_to_target_message_id,
+}) {
+    // These functions are scoped inside `handle_anchor_case` to
+    // to avoid them being used for any other purpose.
+    function duplicate_current_msg_list_with_new_data(data) {
+        const msg_list = new message_list.MessageList({data});
+        msg_list.reading_prevented = message_lists.current.reading_prevented;
+        return msg_list;
+    }
+
+    function select_msg_id(msg_id, select_opts) {
+        message_lists.current.select_id(msg_id, {
+            then_scroll: true,
+            from_scroll: false,
+            ...select_opts,
+        });
+    }
+
+    function select_anchor_using_data(data) {
+        const msg_list = duplicate_current_msg_list_with_new_data(data);
+        message_lists.update_current_message_list(msg_list);
+        // `force_rerender` is required to render the new data.
+        select_msg_id(message_list_data_to_target_message_id(data), {force_rerender: true});
+    }
```

Comment:
I feel like this partial implementation of changing views may be a bit fragile. Do any of these steps need to happen, for example?

```
            // We are navigating to the combined feed from another                                      
            // narrow, so we reset the reading state to allow user to                                   
            // read messages again in the combined feed if user has                                     
            // marked some messages as unread in the last combined                                      
            // feed session and thus prevented reading.                                                 
            message_lists.current.resume_reading();                                                     
            // Reset the collapsed status of messages rows.                                             
            condense.condense_and_collapse(message_lists.current.view.$list.find(".message_row"));      
            message_edit.restore_edit_state_after_message_view_change();                                
            submessage.process_widget_rows_in_list(message_lists.current);                              
            message_feed_top_notices.update_top_of_narrow_notices(msg_list);                            
                                                                                                        
            // We may need to scroll to the selected message after swapping                             
            // the currently displayed center panel to the combined feed.                               
            message_viewport.maybe_scroll_to_selected();                                                
```

I think `message_edit.restore_edit_state_after_message_view_change` might be required; not sure about the rest.

But I'm also worried about whether there will be future bugs where we add some new logic there and it doesn't get properly run here.

Reply:
> I think message_edit.restore_edit_state_after_message_view_change might be required

No, since we are rendering everything again, and that code path includes restoring the edit messages.

>  message_lists.current.resume_reading(); 

We don't need to call this since we want to preserve the current reading state of the user.

> condense.condense_and_collapse(message_lists.current.view.$list.find(".message_row"));
> submessage.process_widget_rows_in_list(message_lists.current); 

Already called in the rerender codepath.


>  message_feed_top_notices.update_top_of_narrow_notices(msg_list);

Don't want to call this since it is a noop since pressing `home/end` key shouldn't impact top of narrow notices.

I understand your consern but using `message_view.show` for this is actually much worse since it is not designed for this use case since its purpose is to show a "new" message list which is not rendered right now and we don't mind losing context for the current message list. While in this case we want to preserve as much context for the current message list that we can and only change selected message.

Reply:
OK, makes sense.